crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -5615,7 +5615,24 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
 
     // Infer the deferred base type of a NewType.
     fn infer_newtype_assignment_deferred(&mut self, arguments: &ast::Arguments) {
-        match self.infer_type_expression(&arguments.args[1]) {
+        let node = &arguments.args[1];
+        let mut base_type = self.infer_type_expression(node);
+        // Detect the `float` and `complex` cases and work around the special case where they
+        // normally refer to unions in type expressions.
+        if let ast::Expr::Name(name) = node {
+            // REVIEWERS: It seems like `infer_expression` here would count as double inference and
+            // panic, but `infer_name_load` is fine because it doesn't store the type? Is this a
+            // principled distinction, or am I hacking around something I shouldn't?
+            let base_value = self.infer_name_load(name);
```

Comment:
This feels very dirty to me, and I'd love suggestions for a better way to do it. The obvious alternative would be hardcoding the names `"float"` and `"complex"`, but that seems excessively brittle?

This is also responsible for the failing test in `new_types.md`, because the lookup fails for a forward reference in a stub file.